In [1]:
import os
import sqlite3
import sys

sys.path.append("..")

import nivapy3 as nivapy
import numpy as np
import pandas as pd
import utils

# Tiltaksovervakingen: opsjon for kvalitetskontroll av analysedata
# Eurofins 2021 Q1

## Notebook 1: Initial exploration and data cleaning

## 1. Initial data check

**Note:** This file uses an updated export from Vannmiljø reflecting corrections made to two station codes in August 2021 - see e-mail from Kjetil received 27.05.2021 at 17:15 for details.

Initial screening of the Eurofins Excel file shows the following:

 * There is one sample with pH < 1, which seems unlikely
 
 * One value for conductivity is given as `.109`. Based on the number formatting and values for other entries in this column, I think this might be `1.09`?
 
 * The column for alkalinity uses `,` as the decimal separator for LOD values. I have replaced all occurrences of `,` as a decimal separator with `.`
 
 * The columns `Vannmengde` and `Temp (C)` are missing from the latest spreadsheet (but have been included in previous rounds). For the QC process described here, `Vannmengde` is not used as it contains qualitative data, but `Temp (C)` has previously been considered. I have excluded temperature from the comparison for this round, but if it is of interest it should be included in the final submission

In [2]:
# Choose dataset to process
lab = "Eurofins"
year = 2021
qtr = 1
version = 2

## 2. Create SQLite database to store results

Using a database will provide basic checks on data integrity and consistency. For this project, three tables will be sufficient:

 * Station locations and metadata
 * Parameters and units used by VestfoldLAB, Eurofins and Vannmiljø, and conversion factors between these
 * Water chemistry data
 
The code below creates a basic database structure, which will be populated later.

In [3]:
# Create database
fold_path = f"../../output/{lab.lower()}_{year}_q{qtr}_v{version}"
if not os.path.exists(fold_path):
    os.makedirs(fold_path)

db_path = os.path.join(fold_path, "kalk_data.db")
if os.path.exists(db_path):
    os.remove(db_path)
eng = sqlite3.connect(db_path, detect_types=sqlite3.PARSE_DECLTYPES)

# Turn off journal mode for performance
eng.execute("PRAGMA synchronous = OFF")
eng.execute("PRAGMA journal_mode = OFF")

# Create stations table
sql = (
    "CREATE TABLE stations "
    "( "
    "  fylke text NOT NULL, "
    "  vassdrag text NOT NULL, "
    "  station_name text NOT NULL, "
    "  station_number text, "
    "  vannmiljo_code text NOT NULL, "
    "  vannmiljo_name text, "
    "  utm_east real NOT NULL, "
    "  utm_north real NOT NULL, "
    "  utm_zone integer NOT NULL, "
    "  lon real NOT NULL, "
    "  lat real NOT NULL, "
    "  liming_status text NOT NULL, "
    "  comment text, "
    "  PRIMARY KEY (vannmiljo_code) "
    ")"
)
eng.execute(sql)

# Create parameters table
sql = (
    "CREATE TABLE parameters_units "
    "( "
    "  vannmiljo_name text NOT NULL UNIQUE, "
    "  vannmiljo_id text NOT NULL UNIQUE, "
    "  vannmiljo_unit text NOT NULL, "
    "  vestfoldlab_name text NOT NULL UNIQUE, "
    "  vestfoldlab_unit text NOT NULL, "
    "  vestfoldlab_to_vm_conv_fac real NOT NULL, "
    "  eurofins_name text NOT NULL UNIQUE, "
    "  eurofins_unit text NOT NULL, "
    "  eurofins_to_vm_conv_fac real NOT NULL, "
    "  min real NOT NULL, "
    "  max real NOT NULL, "
    "  PRIMARY KEY (vannmiljo_id) "
    ")"
)
eng.execute(sql)

# Create chemistry table
sql = (
    "CREATE TABLE water_chemistry "
    "( "
    "  vannmiljo_code text NOT NULL, "
    "  sample_date datetime NOT NULL, "
    "  lab text NOT NULL, "
    "  period text NOT NULL, "
    "  depth1 real, "
    "  depth2 real, "
    "  parameter text NOT NULL, "
    "  flag text, "
    "  value real NOT NULL, "
    "  unit text NOT NULL, "
    "  PRIMARY KEY (vannmiljo_code, sample_date, depth1, depth2, parameter), "
    "  CONSTRAINT vannmiljo_code_fkey FOREIGN KEY (vannmiljo_code) "
    "      REFERENCES stations (vannmiljo_code) "
    "      ON UPDATE NO ACTION ON DELETE NO ACTION, "
    "  CONSTRAINT parameter_fkey FOREIGN KEY (parameter) "
    "      REFERENCES parameters_units (vannmiljo_name) "
    "      ON UPDATE NO ACTION ON DELETE NO ACTION, "
    "  CONSTRAINT unit_fkey FOREIGN KEY (unit) "
    "      REFERENCES parameters_units (vannmiljo_unit) "
    "      ON UPDATE NO ACTION ON DELETE NO ACTION "
    ")"
)
eng.execute(sql)

## 3. Explore station data

Station details are stored in `../../data/active_stations_2020.xlsx`, which is a tidied version of Øyvind's original file here:

    K:\Prosjekter\langtransporterte forurensninger\Kalk Tiltaksovervåking\12 KS vannkjemi\Vannlokaliteter koordinater_kun aktive stasj 2020.xlsx
    
Note that corrections (e.g. adjusted station co-ordinates) have been made to the tidied file, but not the original on `K:`. **The version in this repository should therefore been used as the "master" copy**.

In [4]:
# Read station data
stn_df = pd.read_excel(r"../../data/active_stations_2020.xlsx", sheet_name="data")
stn_df = nivapy.spatial.utm_to_wgs84_dd(stn_df)

print("The following stations are missing spatial co-ordinates:")
stn_df.query("lat != lat")

The following stations are missing spatial co-ordinates:


Empty DataFrame
Columns: [fylke, vassdrag, station_name, station_number, vannmiljo_code, vannmiljo_name, utm_east, utm_north, utm_zone, liming_status, comment, lat, lon]
Index: []

In [5]:
print("The following stations do not have a code in Vannmiljø:")
stn_df.query("vannmiljo_code != vannmiljo_code")

The following stations do not have a code in Vannmiljø:


Empty DataFrame
Columns: [fylke, vassdrag, station_name, station_number, vannmiljo_code, vannmiljo_name, utm_east, utm_north, utm_zone, liming_status, comment, lat, lon]
Index: []

In [6]:
# Map
stn_map = nivapy.spatial.quickmap(
    stn_df.dropna(subset=["lat"]),
    lat_col="lat",
    lon_col="lon",
    popup="station_name",
    cluster=True,
    kartverket=True,
    aerial_imagery=True,
)

stn_map.save("../../pages/stn_map.html")

stn_map

In [7]:
# Add to database
stn_df.dropna(subset=["vannmiljo_code", "lat"], inplace=True)
stn_df.to_sql(name="stations", con=eng, if_exists="append", index=False)

## 4. Parameters and units of interest

The file `../../data/parameter_unit_mapping.xlsx` provides a lookup between parameter names & units used by the labs and those in Vannmiljø. It also contains plausible ranges (using Vannmiljø units) for each parameter. These ranges have been chosen by using the values already in Vannmiljø as a reference. However, it looks as though some of the data in Vannmiljø might also be spurious, so it would be good to refine these ranges based on domain knowledge, if possible.

**Note:** Concentrations reported as *exactly* zero are likely to be errors, because most (all?) lab methods should report an LOQ instead.

In [8]:
# Read parameter mappings
par_df = utils.get_par_unit_mappings()

# Add to database
par_df.to_sql(name="parameters_units", con=eng, if_exists="append", index=False)

par_df

vannmiljo_name vannmiljo_id vannmiljo_unit  \
0                            Temperatur         TEMP             °C   
1                                    pH           PH     <ubenevnt>   
2                         Konduktivitet         KOND           mS/m   
3                       Total alkalitet          ALK         mmol/l   
4                           Totalfosfor        P-TOT         µg/l P   
5                         Totalnitrogen        N-TOT         µg/l N   
6                                Nitrat        N-NO3         µg/l N   
7          Totalt organisk karbon (TOC)          TOC         mg/l C   
8                    Reaktivt aluminium          RAL        µg/l Al   
9                 Ikke-labilt aluminium         ILAL        µg/l Al   
10                     Labilt aluminium          LAL        µg/l Al   
11                               Klorid           CL           mg/l   
12                               Sulfat          SO4           mg/l   
13                              Kalsium           CA           mg/l   
14                               Kalium            K           mg/l   
15                            Magnesium           MG           mg/l   
16                              Natrium           NA           mg/l   
17                       Totalt silikat         SIO2        µg/l Si   
18  Syrenøytraliserende kapasitet (ANC)          ANC         µekv/l   

   vestfoldlab_name vestfoldlab_unit  vestfoldlab_to_vm_conv_fac  \
0              Temp               °C                    1.000000   
1                pH              enh                    1.000000   
2              Kond             mS/m                    1.000000   
3               Alk           mmol/l                    1.000000   
4             Tot-P             µg/l                    1.000000   
5             Tot-N             µg/l                    1.000000   
6               NO3             µg/l                    1.000000   
7               TOC             mg/l                    1.000000   
8               RAl             µg/l                    1.000000   
9              ILAl             µg/l                    1.000000   
10              LAl             µg/l                    1.000000   
11               Cl             mg/l                    1.000000   
12              SO4             mg/l                    1.000000   
13               Ca             mg/l                    1.000000   
14                K             mg/l                    1.000000   
15               Mg             mg/l                    1.000000   
16               Na             mg/l                    1.000000   
17             SIO2             mg/l                  467.543276   
18              ANC           µekv/l                    1.000000   

   eurofins_name eurofins_unit  eurofins_to_vm_conv_fac   min   max  
0           Temp            °C                 1.000000   -10    30  
1             pH           enh                 1.000000     1    10  
2           Kond          ms/m                 1.000000     0   100  
3            Alk        mmol/l                 1.000000     0     2  
4          Tot-P          µg/l                 1.000000     0   500  
5          Tot-N          µg/l                 1.000000     0  4000  
6            NO3          µg/l                 1.000000     0  2000  
7            TOC          mg/l                 1.000000     0   100  
8            RAl          µg/l                 1.000000     0   500  
9           ILAl          µg/l                 1.000000     0   500  
10           LAl          µg/l                 1.000000     0   500  
11            Cl          mg/l                 1.000000     0   100  
12           SO4          mg/l                 1.000000     0    20  
13            Ca          mg/l                 1.000000     0   500  
14             K          mg/l                 1.000000     0    10  
15            Mg          mg/l                 1.000000     0   100  
16            Na          mg/l                 1.0000

## 5. Historic data from Vannmiljø

The Vannmiljø dataset is large and reading from Excel is slow; the code below takes a couple of minutes to run.

Note from the output below that **there are more than 1600 "duplicated" samples in the Vannmiljø dataset** i.e. where the station code, sample date, sample depth, lab and parameter name are all the same, but a different value is reported. It would be helpful to know why these duplicates were collected e.g. are these reanalysis values, where only one of the duplicates should be used, or are they genuine (in which case should they be averaged or kept separate?). **For the moment, I will ignore these values**.

In [9]:
# Read historic data from Vannmiljø
his_df = utils.read_historic_data(
    r"../../data/vannmiljo_export_2012-19_2021-08-16.xlsx"
)

# Tidy lab names for clarity
his_df["lab"].replace(
    {"NIVA": "NIVA (historic)", "VestfoldLAB AS": "VestfoldLAB (historic)"},
    inplace=True,
)

# Check duplicates
key_cols = [
    "vannmiljo_code",
    "sample_date",
    "lab",
    "depth1",
    "depth2",
    "par_unit",
]
his_dup_df = his_df[
    his_df.duplicated(
        key_cols,
        keep=False,
    )
].sort_values(key_cols)
his_dup_df.to_csv(
    r"../../output/vannmiljo_historic/vannmiljo_duplicates.csv", index=False
)

## Average duplicates
# his_df = his_df.groupby(key_cols).aggregate(
#    {
#        "flag": "first",
#        "value": "mean",
#    }
# ).reset_index()

# Remove all duplicates
his_df.drop_duplicates(subset=key_cols, keep=False, inplace=True)

# Add label for data period
his_df["period"] = "historic"

# Print summary
n_stns = len(his_df["vannmiljo_code"].unique())
print(f"The number of unique stations with data is: {n_stns}.\n")

print(
    f"There are {len(his_dup_df)} duplicated records (same station_code-date-depth-parameter, but different value).\n"
    "These will be averaged or ignored (depending on the settings above).\n"
)

his_df.head()

The number of unique stations with data is: 211.

There are 1620 duplicated records (same station_code-date-depth-parameter, but different value).
These will be averaged or ignored (depending on the settings above).



vannmiljo_code sample_date              lab  depth1  depth2       par_unit  \
0      027-28435  2012-01-02  NIVA (historic)     0.0     0.0     ALK_mmol/l   
1      027-28435  2012-01-02  NIVA (historic)     0.0     0.0        CA_mg/l   
2      027-28435  2012-01-02  NIVA (historic)     0.0     0.0   ILAL_µg/l Al   
3      027-28435  2012-01-02  NIVA (historic)     0.0     0.0      KOND_mS/m   
4      027-28435  2012-01-02  NIVA (historic)     0.0     0.0  PH_<ubenevnt>   

  flag  value    period  
0    =   0.06  historic  
1    =   1.23  historic  
2    =  18.00  historic  
3    =   3.80  historic  
4    =   6.24  historic

## 6. New data from labs

The code below reads the Excel template provided by Eurofins and reformats it to the same structure (parameter names, units etc.) as the data in Vannmiljø.

In [10]:
# Read new data
new_df = utils.read_data_template(
    f"../../data/{lab.lower()}_data_{year}_q{qtr}_v{version}.xlsx",
    sheet_name="results",
    lab=lab,
)

# Check duplicates
new_dup_df = new_df[
    new_df.duplicated(
        key_cols,
        keep=False,
    )
].sort_values(key_cols)
dup_csv = os.path.join(fold_path, f"{lab.lower()}_{year}_q{qtr}_v{version}_duplicates.csv")
new_dup_df.to_csv(dup_csv, index=False)

## Average duplicates
# new_df = (
#    new_df.groupby(key_cols)
#    .aggregate(
#        {
#            "flag": "first",
#            "value": "mean",
#        }
#    )
#    .reset_index()
# )

# Remove all duplicates
new_df.drop_duplicates(subset=key_cols, keep=False, inplace=True)

# Add label for data period
new_df["period"] = "new"

print(
    f"\nThere are {len(new_dup_df)} duplicated records (same station_code-date-depth-parameter, but different value).\n"
    "These will be averaged or ignored (depending on the settings above).\n"
)

new_df.head()


The following location IDs have inconsistent names within this template:

The following location names have multiple IDs within this template:

The following samples have nitrate greater than total nitrogen:
     vannmiljo_code         sample_date depth1 depth2  NO3_µg/l  Tot-N_µg/l
88        022-90731 2021-02-03 07:00:50      0      0     510.0         500
431       027-38543 2021-02-02 07:00:34      0      0     430.0         420
443       030-58776 2021-01-13 07:15:13      0      0     350.0         330
446       030-58776 2021-02-09 07:10:34      0      0     400.0         390
586       031-58843 2021-01-07 07:00:30      0      0     180.0         170
587       031-58843 2021-01-13 07:15:13      0      0     170.0         160
590       031-58843 2021-02-10 07:15:47      0      0     210.0         200
602       031-38523 2021-02-10 07:15:47      0      0     290.0         280
701       038-58854 2021-01-05 07:00:45      0      0     530.0         520
703       038-58854 2021-02-02 

vannmiljo_code         sample_date       lab  depth1  depth2       par_unit  \
0      023-45835 2021-01-06 07:10:46  Eurofins     0.0     0.0  PH_<ubenevnt>   
1      023-45835 2021-02-03 07:00:13  Eurofins     0.0     0.0  PH_<ubenevnt>   
2      023-45835 2021-03-02 07:00:37  Eurofins     0.0     0.0  PH_<ubenevnt>   
3      023-62175 2021-01-06 07:10:46  Eurofins     0.0     0.0  PH_<ubenevnt>   
4      023-62175 2021-01-25 09:00:07  Eurofins     0.0     0.0  PH_<ubenevnt>   

  flag  value period  
0  nan    5.5    new  
1  nan    5.7    new  
2  nan    5.5    new  
3  nan    6.2    new  
4  nan    6.1    new

## 7 . Combine

Combine the `historic` and `new` datasets into a single dataframe in "long" format.

In [11]:
# Combine
df = pd.concat([his_df, new_df], axis="rows")

# Separate par and unit
df[["parameter", "unit"]] = df["par_unit"].str.split("_", n=1, expand=True)
del df["par_unit"]

df.reset_index(drop=True, inplace=True)

df.head()

vannmiljo_code sample_date              lab  depth1  depth2 flag  value  \
0      027-28435  2012-01-02  NIVA (historic)     0.0     0.0    =   0.06   
1      027-28435  2012-01-02  NIVA (historic)     0.0     0.0    =   1.23   
2      027-28435  2012-01-02  NIVA (historic)     0.0     0.0    =  18.00   
3      027-28435  2012-01-02  NIVA (historic)     0.0     0.0    =   3.80   
4      027-28435  2012-01-02  NIVA (historic)     0.0     0.0    =   6.24   

     period parameter        unit  
0  historic       ALK      mmol/l  
1  historic        CA        mg/l  
2  historic      ILAL     µg/l Al  
3  historic      KOND        mS/m  
4  historic        PH  <ubenevnt>

In [12]:
# Apply correction to historic SIO2
df["value"] = np.where(
    (df["lab"] == "VestfoldLAB (historic)") & (df["parameter"] == "SIO2"),
    df["value"] * 467.5432,
    df["value"],
)

# Reclassify (nitrate + nitrite) to nitrate
df["parameter"].replace({"N-SNOX": "N-NO3"}, inplace=True)

## 8. Check data ranges

A simple method for preliminary quality control is to check whether parameter values are within sensible ranges (as defined in the `parameters_units` table; see Section 4 above). I believe this screening should be implemented differently for the `historic` (i.e. Vannmiljø) and `new` datsets, as follows:

 * For the `historic` data in Vannmiljø, values outside the plausible ranges should be **removed from the dataset entirely**. This is because we intend to use the Vannmiljø data as a reference against which new values will be compared, so it is important the dataset does not contain anything too strange. Ideally, the reference dataset should be carefully manually curated to ensure it is as good as possible, but I'm not sure we have the resouces in this project to thoroughly quality assess the data *already* in Vannmiljø. Dealing with any obvious issues is a good start, though
 
 * For the `new` data, values outside the plausible ranges should be highlighted and checked with the reporting lab
 
**Note:** At present, my code will remove any concentration values of exactly zero from the historic dataset. **Check with Øyvind whether this is too strict**.

In [13]:
# Check ranges
df = utils.check_data_ranges(df)


Checking data ranges for the 'historic' period.
    KOND: Maximum value of 309.00 is greater than or equal to upper limit (100.00).
    LAL: Minimum value of 0.00 is less than or equal to lower limit (0.00).
    SO4: Minimum value of 0.00 is less than or equal to lower limit (0.00).
    CA: Minimum value of 0.00 is less than or equal to lower limit (0.00).
    SIO2: Maximum value of 51897.30 is greater than or equal to upper limit (7000.00).

Checking data ranges for the 'new' period.
    N-NO3: Maximum value of 2300.00 is greater than or equal to upper limit (2000.00).

Dropping problem rows from historic data.
    Dropping rows for KOND.
    Dropping rows for LAL.
    Dropping rows for SO4.
    Dropping rows for CA.
    Dropping rows for SIO2.


In [14]:
# Add to database
df.to_sql(
    name="water_chemistry",
    con=eng,
    if_exists="append",
    index=False,
    method="multi",
    chunksize=1000,
)

In [15]:
eng.close()

## 9. Version 1 summary

Points to note from the initial data exploration:

### 9.1. Data formatting and missing data

 * There is one sample with pH `<1`, which seems unlikely
 
 * One value for conductivity is given as `.109`. Based on the number formatting and values for other entries in this column, I think this might be `1.09`?
 
 * The column for alkalinity uses `,` as the decimal separator for LOD values. I have replaced all occurrences of `,` as a decimal separator with `.`
 
 * The columns `Vannmengde` and `Temp (C)` are missing from the latest spreadsheet (but have been included in previous rounds). For the QC process described here, `Vannmengde` is not used as it contains qualitative data, but `Temp (C)` has previously been considered. I have excluded temperature from the comparison for this round, but if it is of interest it should be included in the final submission
 
### 9.2. Site IDs

**Note:** This check is a new addition for this round of screening. I have added it because in the previous round there were some issues due to the same location being assigned multiple station IDs (most likely caused by the "autoincrement" feature in Excel). I have previously checked that each record in the template has a valid station ID, but this check goes further by ensuring that IDs and site names are used consistently within each file. As this test is new, **it is possible that such errors have occurred previously but not been detected**.

The following location IDs have inconsistent names within this template:

     025-45779 ['Litlåna oppstrøms doserer (5)']   ==>   ['Litlåna' 'Litlåna oppstrøms doserer' 'Logen nedstrøms doserer']
     025-81555 ['Stakkeland bru (Rv. 42) (9)']   ==>   ['Stakkeland bru (Rv. 42)' 'Stakkeland bru (Rv. 42) ']
     020-45060 ['Vindsland oppstrøms']   ==>   ['Vindsland nedstrøms' 'Vindsland oppstrøms']
     020-40521 ['Tveitvatn utløp (4)']   ==>   ['Tveitvatn utløp' 'Tverrbekken']
     038-58866 ['Litleelva nedstrøms doserer (17)']   ==>   ['Litleelva nedstrøms doserer' 'Litleåna nedstrøms doserer']
     038-58747 ['Litleelva oppstrøms doserer (16)']   ==>   ['Litleelva oppstrøms doserer' 'Litleåna oppstrøms doserer']
     063-62170 ['Mysterelvi (1b)']   ==>   ['Mysterelvi' 'Mysterøyri']
     067-82876 ['Daleelva nedenfor kraftverk (3)']   ==>   ['Daleelva nedenfor kraftverk' 'Dåsåna nedstrøms doserer']
     064-82801 ['Krossdalselva (5)']   ==>   ['Krossdalselva' 'Krossdalselva ']
     045-58815 ['Børsdalselva oppstrøms doserer (d)']   ==>   ['Børsdalselva oppstrøms doserer' 'Børsdalselva oppstrøms doserer ']
     045-58817 ['Uskedalselva øverst (b)']   ==>   ['Uskedalselva utløp' 'Uskedalselva øverst']
     062-58818 ['Bolstadelva (9)']   ==>   ['Bolstadelva' 'Bolstadelva ']
     067-58824 ['Botnabekken (13)']   ==>   ['Botnabekken' 'Botnabekken ']

The following location names have multiple IDs within this template:

     Vindsland nedstrøms ['020-45013']   ==>   ['020-45013' '020-45060']
     Tverrbekken []   ==>   ['020-40521' '063-58808']
     Uskedalselva utløp []   ==>   ['045-58816' '045-58817']
 
### 9.3. Duplicates

 * There are **204 duplicated records** in the Eurofins data submission, where same station_code-date-depth-parameter has been reported more than once (see [here](https://github.com/NIVANorge/tiltaksovervakingen/blob/master/output/eurofins_2021_q1_duplicates.csv) for a list). It looks as though these are genuine duplicates (i.e. these samples have been analysed more than once), because they have different `Labreferanse` codes in the spreadsheet. Are these reanalysis values (in which cases only the best/most recent value should probably be included), or do they have another significance? **Check with Eurofins**
 
### 9.4. Range checks

**Note:** Some of the points identified below are explored more thoroughly in the time series analysis.

 * Values for nitrate > 2000 ug/l seem very high?
 
 * Values of pH reported as `<1` seem very low?
 
The following samples have nitrate greater than total nitrogen:

         vannmiljo_code         sample_date depth1 depth2  NO3_µg/l  Tot-N_µg/l
    88        022-90731 2021-02-03 07:00:50      0      0     510.0         500
    432       027-38543 2021-02-02 07:00:34      0      0     430.0         420
    444       030-58776 2021-01-13 07:15:13      0      0     350.0         330
    447       030-58776 2021-02-09 07:10:34      0      0     400.0         390
    587       031-58843 2021-01-07 07:00:30      0      0     180.0         170
    588       031-58843 2021-01-13 07:15:13      0      0     170.0         160
    591       031-58843 2021-02-10 07:15:47      0      0     210.0         200
    603       031-38523 2021-02-10 07:15:47      0      0     290.0         280
    702       038-58854 2021-01-05 07:00:45      0      0     530.0         520
    704       038-58854 2021-02-02 07:00:34      0      0     480.0         460
    705       038-58854 2021-02-09 07:10:07      0      0     600.0         580
    928       038-58868 2021-02-02 07:00:34      0      0     350.0         340
    967       063-62170 2021-02-19 07:00:10      0      0     370.0         350
    1136      067-82874 2021-02-02 07:00:34      0      0     260.0         250
    1217      064-82800 2021-02-02 07:00:43      0      0     270.0         250
    1306      045-58816 2021-02-03 07:00:50      0      0     680.0         640
    
## 10. Version 2 summary (02.02.2022)

I have re-run this notebook with the cleaned "v2" dataset from Marianne (received 21.01.2022 at 15.21).

 * Alk-E at 030-58838 (Frafjordelva) on 12.01.2021 has a value of `103'8`. I assume this is a new typo (maybe `103.8`?). **Check with Marianne**
 
 * There are 36 duplicated records (presumably flood samples etc.) 

The following samples have nitrate greater than total nitrogen:

         vannmiljo_code         sample_date depth1 depth2  NO3_µg/l  Tot-N_µg/l
    88        022-90731 2021-02-03 07:00:50      0      0     510.0         500
    431       027-38543 2021-02-02 07:00:34      0      0     430.0         420
    443       030-58776 2021-01-13 07:15:13      0      0     350.0         330
    446       030-58776 2021-02-09 07:10:34      0      0     400.0         390
    586       031-58843 2021-01-07 07:00:30      0      0     180.0         170
    587       031-58843 2021-01-13 07:15:13      0      0     170.0         160
    590       031-58843 2021-02-10 07:15:47      0      0     210.0         200
    602       031-38523 2021-02-10 07:15:47      0      0     290.0         280
    701       038-58854 2021-01-05 07:00:45      0      0     530.0         520
    703       038-58854 2021-02-02 07:00:34      0      0     480.0         460
    704       038-58854 2021-02-09 07:10:07      0      0     600.0         580
    927       038-58868 2021-02-02 07:00:34      0      0     350.0         340
    966       063-58804 2021-02-19 07:00:10      0      0     370.0         350
    1136      067-82874 2021-02-02 07:00:34      0      0     260.0         250
    1217      064-82800 2021-02-02 07:00:43      0      0     270.0         250
    1306      045-58816 2021-02-03 07:00:50      0      0     680.0         640